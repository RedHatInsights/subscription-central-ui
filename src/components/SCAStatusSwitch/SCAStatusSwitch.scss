.sca-status-error-icon {
  margin-top: 2px;
  margin-bottom: -2px;
}

.sca-status-error-msg {
  font-size: 12px;
  line-height: 1.4em;
  width: calc(100% - 23px);
  float: right;
}

.sca-status-spinner {
  margin: 6px 0 -6px 5px;
}
