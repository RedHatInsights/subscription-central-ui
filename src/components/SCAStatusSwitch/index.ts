import SCAStatusSwitch from './SCAStatusSwitch';

export default SCAStatusSwitch;
