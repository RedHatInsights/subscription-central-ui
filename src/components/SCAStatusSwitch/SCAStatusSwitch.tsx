import React, { FC } from 'react';
import { Spinner, Switch } from '@patternfly/react-core';
import ExclamationCircleIcon from '@patternfly/react-icons/dist/js/icons/exclamation-circle-icon';
import useUpdateManifestSCAStatus from '../../hooks/useUpdateManifestSCAStatus';
import './SCAStatusSwitch.scss';

interface SCAStatusSwitchProps {
  scaStatus: string;
  uuid: string;
}

const SCAStatusSwitch: FC<SCAStatusSwitchProps> = ({ scaStatus, uuid }) => {
  const isChecked = scaStatus.toLowerCase() === 'enabled';

  const { data, mutate, isLoading, isError, isSuccess } = useUpdateManifestSCAStatus();

  const failedUpdatingSCAStatus = isError === true || (isSuccess && !data);

  const updateManifestSCAStatus = (uuid: string, currentSCAStatus: string) => {
    const newSCAStatus = currentSCAStatus.toLowerCase() === 'enabled' ? 'disabled' : 'enabled';
    mutate({ uuid, newSCAStatus });
  };

  const handleChange = () => {
    updateManifestSCAStatus(uuid, scaStatus);
  };

  const ErrorState = () => (
    <>
      <ExclamationCircleIcon
        color="var(--pf-global--danger-color--100)"
        className="sca-status-error-icon"
      />
      <span className="sca-status-error-msg">
        Something went wrong. Please refresh the page and try again.
      </span>
    </>
  );

  const LoadingState = () => <Spinner size="lg" className="sca-status-spinner" />;

  const DisallowedState = () => <p>N/A</p>;

  const SCASwitch = () => (
    <Switch
      id={`sca-status-switch-${uuid}`}
      data-testid="sca-status-switch"
      aria-label={`SCA Status for this Manifest is ${scaStatus}`}
      isChecked={isChecked}
      onChange={handleChange}
      label="Enabled"
      labelOff="Disabled"
    />
  );

  if (failedUpdatingSCAStatus === true) {
    return <ErrorState />;
  } else if (isLoading === true) {
    return <LoadingState />;
  } else if (scaStatus === 'disallowed') {
    return <DisallowedState />;
  } else {
    return <SCASwitch />;
  }
};

export default SCAStatusSwitch;
