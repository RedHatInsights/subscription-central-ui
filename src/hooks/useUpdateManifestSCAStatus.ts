import { useMutation, useQueryClient, UseMutationResult, QueryClient } from 'react-query';
import { ManifestEntry } from './useSatelliteManifests';
import { ManifestEntitlementsData } from './useManifestEntitlements';
import { useToken } from '../utilities/platformServices';

interface UpdateManifestSCAStatusParams {
  uuid: string;
  newSCAStatus: string;
}

interface UpdateManifestSCAStatus {
  success: boolean;
  status: number;
}

const updateManifestSCAStatus =
  (jwtToken: Promise<string>) =>
  async (data: UpdateManifestSCAStatusParams): Promise<void | UpdateManifestSCAStatus> => {
    const { uuid, newSCAStatus } = data;

    const requestData = {
      simpleContentAccess: newSCAStatus
    };

    return fetch(`/api/rhsm/v2/manifests/${uuid}`, {
      method: 'PUT',
      headers: { Authorization: `Bearer ${await jwtToken}` },
      body: JSON.stringify(requestData)
    })
      .then((response) => {
        if (response.status !== 204) {
          throw new Error(
            `Status Code ${response.status}.  Error updating SCA status: ${response.statusText}.  `
          );
        }
        return { success: true, status: response.status };
      })
      .catch((e) => {
        console.error(e);
      });
  };

const updateManifestSCAQueryData = (
  queryClient: QueryClient,
  uuid: string,
  newSCAStatus: string
): void => {
  const oldManifestEntryData: ManifestEntry[] = queryClient.getQueryData('manifests');

  const newManifestEntryData: ManifestEntry[] = oldManifestEntryData.map((manifestEntry) => {
    if (manifestEntry.uuid === uuid) {
      const updatedManifestEntry = { ...manifestEntry };
      updatedManifestEntry.simpleContentAccess = newSCAStatus;
      return updatedManifestEntry;
    } else {
      return manifestEntry;
    }
  });

  queryClient.setQueryData('manifests', newManifestEntryData);
};

const updateManifestEntitlementsData = (
  queryClient: QueryClient,
  uuid: string,
  newSCAStatus: string
): void => {
  const oldEntitlementsData: ManifestEntitlementsData = queryClient.getQueryData([
    'manifestEntitlements',
    uuid
  ]);

  if (!oldEntitlementsData) return;

  const newEntitlementsData = {
    body: { ...oldEntitlementsData.body, simpleContentAccess: newSCAStatus }
  };

  queryClient.setQueryData(['manifestEntitlements', uuid], newEntitlementsData);
};

const useUpdateManifestSCAStatus = (): UseMutationResult<
  void | UpdateManifestSCAStatus,
  unknown
> => {
  const queryClient = useQueryClient();
  const jwtToken = useToken();
  return useMutation(
    (updateManifestSCAStatusParams: UpdateManifestSCAStatusParams) =>
      updateManifestSCAStatus(jwtToken)(updateManifestSCAStatusParams),
    {
      onSuccess: (data, updateManifestSCAStatusParams) => {
        if (!data) return;

        const { uuid, newSCAStatus } = updateManifestSCAStatusParams;

        updateManifestSCAQueryData(queryClient, uuid, newSCAStatus);

        updateManifestEntitlementsData(queryClient, uuid, newSCAStatus);
      }
    }
  );
};

export {
  UpdateManifestSCAStatusParams,
  updateManifestSCAStatus,
  useUpdateManifestSCAStatus as default
};
