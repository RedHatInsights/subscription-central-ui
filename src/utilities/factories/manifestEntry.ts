import { faker } from '@faker-js/faker';
import { Factory } from 'fishery';
import { ManifestEntry } from '../../hooks/useSatelliteManifests';

export default Factory.define<ManifestEntry>(() => ({
  entitlementQuantity: faker.number.int(100),
  name: faker.lorem.word(10),
  type: 'Satellite',
  url: faker.internet.url(),
  uuid: faker.string.uuid(),
  version: '6.9',
  simpleContentAccess: 'enabled'
}));
