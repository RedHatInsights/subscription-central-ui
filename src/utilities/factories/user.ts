import { Factory } from 'fishery';
import { User } from '../../hooks/useUser';

export default Factory.define<User>(() => ({
  canReadManifests: true,
  canWriteManifests: true,
  isOrgAdmin: true,
  isSCACapable: true,
  isEntitled: true
}));
