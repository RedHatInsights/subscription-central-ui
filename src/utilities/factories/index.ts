import manifestEntry from './manifestEntry';
import user from './user';

export default { manifestEntry, user };
