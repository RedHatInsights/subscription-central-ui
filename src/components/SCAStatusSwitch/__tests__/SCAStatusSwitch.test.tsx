import React from 'react';
import { render, fireEvent, getByTestId, waitFor } from '@testing-library/react';
import { QueryClient } from 'react-query';
import SCAStatusSwitch from '../SCAStatusSwitch';
import useUpdateManifestSCAStatus from '../../../hooks/useUpdateManifestSCAStatus';
import factories from '../../../utilities/factories';
import { get, def } from 'bdd-lazy-var';

jest.mock('../../../hooks/useUser');
jest.mock('../../../hooks/useUpdateManifestSCAStatus');

const queryClient = new QueryClient();

describe('Satellite Manifest Panel', () => {
  describe('SCAStatusSwitch', () => {
    def('scaCapable', () => true);
    def('canReadManifests', () => true);
    def('canWriteManifests', () => true);
    def('user', () => {
      return factories.user.build({
        isSCACapable: get('scaCapable'),
        canWriteManifests: get('canWriteManifests'),
        canReadManifests: get('canReadManifests')
      });
    });
    def('props', () => {
      return {
        data: get('data'),
        isError: get('Error'),
        user: get('user')
      };
    });

    const props = {
      scaStatus: 'enabled',
      uuid: 'abc123'
    };
    beforeEach(() => {
      queryClient.setQueryData('user', get('user'));
    });

    it('renders with a spinner when loading', () => {
      (useUpdateManifestSCAStatus as jest.Mock).mockImplementation(() => ({
        isLoading: true
      }));

      const container = render(<SCAStatusSwitch user={get('user')} {...props} />);
      expect(container).toHaveLoader();
    });

    it('renders with an error message when an error occurs', () => {
      (useUpdateManifestSCAStatus as jest.Mock).mockImplementation(() => ({
        isError: true
      }));

      const { getByText } = render(<SCAStatusSwitch user={get('user')} {...props} />);

      expect(
        getByText('Something went wrong. Please refresh the page and try again.')
      ).toBeInTheDocument();
    });

    it('renders with an error message when the API response does not have 204 status and no data is returned', () => {
      (useUpdateManifestSCAStatus as jest.Mock).mockImplementation(() => ({
        isError: false,
        isLoading: false,
        isSuccess: true,
        data: null
      }));

      const { getByText } = render(<SCAStatusSwitch user={get('user')} {...props} />);

      expect(
        getByText('Something went wrong. Please refresh the page and try again.')
      ).toBeInTheDocument();
    });

    it('renders successfully as enabled when success is returned across', () => {
      (useUpdateManifestSCAStatus as jest.Mock).mockImplementation(() => ({
        isError: false,
        isSuccess: true,
        isLoading: false,
        data: { success: true, status: 204 }
      }));

      const { getByLabelText } = render(<SCAStatusSwitch user={get('user')} {...props} />);
      expect(getByLabelText('SCA Status for this Manifest is enabled')).toBeChecked();
    });

    it('renders successfully as disabled when success is returned across', () => {
      (useUpdateManifestSCAStatus as jest.Mock).mockImplementation(() => ({
        isError: false,
        isSuccess: true,
        isLoading: false,
        data: { success: true, status: 204 }
      }));
      const props = {
        scaStatus: 'disabled',
        uuid: 'abc123'
      };

      const { getByLabelText } = render(<SCAStatusSwitch user={get('user')} {...props} />);
      expect(getByLabelText('SCA Status for this Manifest is disabled')).not.toBeChecked();
    });

    it('renders as disallowed when disallowed is passed as scaStatus', () => {
      (useUpdateManifestSCAStatus as jest.Mock).mockImplementation(() => ({
        isError: false,
        isSuccess: false,
        isLoading: false
      }));
      const props = {
        scaStatus: 'disallowed',
        uuid: 'abc123'
      };

      const { getByText } = render(<SCAStatusSwitch user={get('user')} {...props} />);
      expect(getByText('N/A')).toBeInTheDocument();
    });

    it('fires the handleChange when clicked', async () => {
      (useUpdateManifestSCAStatus as jest.Mock).mockImplementation(() => ({
        isError: false,
        isSuccess: false,
        isLoading: false,
        mutate: (): unknown => null
      }));

      const { container } = render(<SCAStatusSwitch user={get('user')} {...props} />);
      fireEvent.click(getByTestId(container, 'sca-status-switch'));
      await waitFor(() => expect(useUpdateManifestSCAStatus).toHaveBeenCalledTimes(1));
    });

    describe('when user does not have write permission', () => {
      def('canWriteManifests', () => false);

      it('displays disabled state of the SCA toggle ', () => {
        (useUpdateManifestSCAStatus as jest.Mock).mockImplementation(() => ({
          isError: false,
          isSuccess: false,
          isLoading: false
        }));

        const { container } = render(<SCAStatusSwitch user={get('user')} {...props} />);

        waitFor(() => expect(container.querySelector('.pf-v5-c-switch__toggle')).toBeDisabled());
      });
    });
  });
});
