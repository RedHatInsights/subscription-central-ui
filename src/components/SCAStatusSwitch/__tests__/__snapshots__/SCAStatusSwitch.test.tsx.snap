// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SCAStatusSwitch renders as disallowed when disallowed is passed as scaStatus 1`] = `
<div>
  <p>
    N/A
  </p>
</div>
`;

exports[`SCAStatusSwitch renders successfully as disabled when success is returned across 1`] = `
<div>
  <label
    class="pf-c-switch"
    data-ouia-component-id="OUIA-Generated-Switch-2"
    data-ouia-component-type="PF4/Switch"
    data-ouia-safe="true"
    for="sca-status-switch-abc123"
  >
    <input
      aria-label="SCA Status for this Manifest is disabled"
      class="pf-c-switch__input"
      data-testid="sca-status-switch"
      id="sca-status-switch-abc123"
      type="checkbox"
    />
    <span
      class="pf-c-switch__toggle"
    />
    <span
      aria-hidden="true"
      class="pf-c-switch__label pf-m-on"
    >
      Enabled
    </span>
    <span
      aria-hidden="true"
      class="pf-c-switch__label pf-m-off"
    >
      Disabled
    </span>
  </label>
</div>
`;

exports[`SCAStatusSwitch renders successfully as enabled when success is returned across 1`] = `
<div>
  <label
    class="pf-c-switch"
    data-ouia-component-id="OUIA-Generated-Switch-1"
    data-ouia-component-type="PF4/Switch"
    data-ouia-safe="true"
    for="sca-status-switch-abc123"
  >
    <input
      aria-label="SCA Status for this Manifest is enabled"
      checked=""
      class="pf-c-switch__input"
      data-testid="sca-status-switch"
      id="sca-status-switch-abc123"
      type="checkbox"
    />
    <span
      class="pf-c-switch__toggle"
    />
    <span
      aria-hidden="true"
      class="pf-c-switch__label pf-m-on"
    >
      Enabled
    </span>
    <span
      aria-hidden="true"
      class="pf-c-switch__label pf-m-off"
    >
      Disabled
    </span>
  </label>
</div>
`;

exports[`SCAStatusSwitch renders with a spinner when loading 1`] = `
<div>
  <span
    aria-valuetext="Loading..."
    class="pf-c-spinner pf-m-lg sca-status-spinner"
    role="progressbar"
  >
    <span
      class="pf-c-spinner__clipper"
    />
    <span
      class="pf-c-spinner__lead-ball"
    />
    <span
      class="pf-c-spinner__tail-ball"
    />
  </span>
</div>
`;

exports[`SCAStatusSwitch renders with an error message when an error occurs 1`] = `
<div>
  <svg
    aria-hidden="true"
    class="sca-status-error-icon"
    fill="var(--pf-global--danger-color--100)"
    height="1em"
    role="img"
    style="vertical-align: -0.125em;"
    viewBox="0 0 512 512"
    width="1em"
  >
    <path
      d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
    />
  </svg>
  <span
    class="sca-status-error-msg"
  >
    Something went wrong. Please refresh the page and try again.
  </span>
</div>
`;

exports[`SCAStatusSwitch renders with an error message when the API response does not have 204 status and no data is returned 1`] = `
<div>
  <svg
    aria-hidden="true"
    class="sca-status-error-icon"
    fill="var(--pf-global--danger-color--100)"
    height="1em"
    role="img"
    style="vertical-align: -0.125em;"
    viewBox="0 0 512 512"
    width="1em"
  >
    <path
      d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
    />
  </svg>
  <span
    class="sca-status-error-msg"
  >
    Something went wrong. Please refresh the page and try again.
  </span>
</div>
`;
